staging/src/k8s.io/mount-utils/resizefs_linux.go

```diff
@@ -84,3 +86,137 @@ func (resizefs *ResizeFs) xfsResize(deviceMountPath string) (bool, error) {
+	case "xfs":
+		blockSize, fsSize, err = resizefs.getXFSSize(deviceMountPath)
+		klog.V(5).Infof("Xfs size: filesystem size=%d, block size=%d, err=%v", fsSize, blockSize, err)
+	default:
+		klog.Errorf("Not able to parse given filesystem info. fsType: %s, will not resize", format)
+		return false, fmt.Errorf("Could not parse fs info on given filesystem format: %s. Supported fs types are: xfs, ext3, ext4", format)
+	}
+	if err != nil {
+		return false, err
+	}
+	// Tolerate one block difference, just in case of rounding errors somewhere.
+	klog.V(5).Infof("Volume %s: device size=%d, filesystem size=%d, block size=%d", devicePath, deviceSize, fsSize, blockSize)
+	if deviceSize <= fsSize+blockSize {
+		return false, nil
+	}
+	return true, nil
+}
+func (resizefs *ResizeFs) getDeviceSize(devicePath string) (uint64, error) {
+	output, err := resizefs.exec.Command("blockdev", "--getsize64", devicePath).CombinedOutput()
+	outStr := strings.TrimSpace(string(output))
+	if err != nil {
+		return 0, fmt.Errorf("failed to read size of device %s: %s: %s", devicePath, err, outStr)
+	}
+	size, err := strconv.ParseUint(outStr, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse size of device %s %s: %s", devicePath, outStr, err)
+	}
+	return size, nil
+}
+
+func (resizefs *ResizeFs) getExtSize(devicePath string) (uint64, uint64, error) {
+	output, err := resizefs.exec.Command("dumpe2fs", "-h", devicePath).CombinedOutput()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to read size of filesystem on %s: %s: %s", devicePath, err, string(output))
+	}
+
+	lines := strings.Split(string(output), "\n")
+	var blockSize, blockCount uint64
+
+	for _, line := range lines {
+		tokens := strings.Split(line, ":")
+		if len(tokens) != 2 {
+			continue
+		}
+		key, value := strings.TrimSpace(tokens[0]), strings.TrimSpace(tokens[1])
+		if key == "Block count" {
+			blockCount, err = strconv.ParseUint(value, 10, 64)
+			if err != nil {
+				return 0, 0, fmt.Errorf("failed to parse block count %s: %s", value, err)
+			}
+		}
+		if key == "Block size" {
+			blockSize, err = strconv.ParseUint(value, 10, 64)
+			if err != nil {
+				return 0, 0, fmt.Errorf("failed to parse block size %s: %s", value, err)
+			}
+		}
+	}
+	if blockSize == 0 {
+		return 0, 0, fmt.Errorf("could not find block size of device %s", devicePath)
+	}
+	if blockCount == 0 {
+		return 0, 0, fmt.Errorf("could not find block count of device %s", devicePath)
+	}
+	return blockSize, blockSize * blockCount, nil
+}
+
+func (resizefs *ResizeFs) getXFSSize(devicePath string) (uint64, uint64, error) {
+	output, err := resizefs.exec.Command("xfs_info", devicePath).CombinedOutput()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to read size of filesystem on %s: %s: %s", devicePath, err, string(output))
+	}
+
+	lines := strings.Split(string(output), "\n")
+	var blockSize, blockCount uint64
+
+	for _, line := range lines {
```

Comment:
cc @sandeen for xfs_info parsing.

Reply:
@sandeen any updates on this?

Reply:
@sandeen, I may miss something here, but IMO `statfs` shows size of the filesystem available for actual files (i.e. without journal and other FS metadata), while xfs_info shows the actual size occupied by the filesystem on a device, incl. journal, inodes, and whatnot.

We need to reliably check if we should call `resize2fs` / `xfs_growfs`, i.e. there is some extra space on the device that the filesystem does not know about.

```
$ df  --output=size /boot
1K-blocks
   999320
$ dumpe2fs /dev/nvme0n1p2
Block count:              262144
Block size:               4096
```
999320 * 1024 = 1023303680
262144 * 4096 = 1073741824

That's quite a difference.

Reply:
It's < 5%, or about 58MB.

So, yes - as I mentioned it won't count filesystem overhead like the journal, or inode tables & bitmaps on ext4.

What is your typical image size, roughly?

On a 2G filesystem image, with default mkfs, filesystem overhead is about 5% for ext4, and < 1% for xfs.  If that is "quite a difference" and outside your threshold, then you may be stuck with the more fragile screenscraping as proposed.

Or, you could  just do:

```
mount
stat -f
if device is more than 10% bigger than free blocks
    try resize2fs/xfs_growfs
    # if you guessed wrong, the resize will be a no-op.
```
Or, just try a growfs after each initial mount, which is going to be quick and harmless if the filesystem is already maximally sized.

That way you could dispense with the screenscraping, which has a reasonable chance of breaking in the future, because admin tool text output is really not an API, and unfortunately there is not any real API to tell you exactly what the last block known to a filesystem is.

Reply:
> just try a growfs after each initial mount, which is going to be quick and harmless if the filesystem is already maximally sized.

How safe is this for xfs and ext3/ext4? Since we will do it with every mount() and containers can move around the cluster rather quickly, we need it to be fast enough (say same as mount()?) and reliable (won't ever harm the FS, even when it gets SIGINT).

> What is your typical image size, roughly?

It starts with 1 GB and goes to hundreds of GB. But I am not sure it's relevant. If user extends the volume by 1% (which can be say 1GB), we'd like the volume to be be expanded.

Reply:
Let's chat on IRC when you get a chance, higher bandwidth than github comments might help.

Reply:
Correction: is go & `ioctl` portable to ARM / PPC / s390x and other weird architectures that Kubernetes supports? If not, do not hesitate to fall back to parsing `xfs_io -c statfs`.

Reply:
If you parse the xfs_io output, the field you want is "geom.datablocks" to get the actual, complete number of all blocks on the filesytem's data device. Note that this is only available when running the xfs_io "statfs" command on an xfs filesystem, non-xfs filesystems will only return the basic statfs(2) information, same as df.